File: app/src/main/java/com/github/nickpakhomov/hotel_viewer/ui/activities/BaseActivity.java
package com.github.nickpakhomov.hotel_viewer.ui.activities;

import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.github.nickpakhomov.hotel_viewer.R;
import com.github.nickpakhomov.hotel_viewer.handler.AnimationHandlerImpl;
import com.github.nickpakhomov.hotel_viewer.handler.FetcherHandlerImpl;

/**
 * Base activity that creates {@link FetcherHandlerImpl} and {@link AnimationHandlerImpl}
 * <br/>
 * Created by Nikolay Pakhomov on 08/09/17.
 */

public class BaseActivity extends AppCompatActivity {
    
    private AnimationHandlerImpl mAnimationHandler;
    FetcherHandlerImpl mFetcherHandler;
    
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        if (this instanceof MainActivity)
            inflater.inflate(R.menu.menu_main, menu);
        else if (this instanceof DetailActivity)
            inflater.inflate(R.menu.menu_detailed, menu);
        
        final MenuItem item = menu.findItem(R.id.action_refresh);
        
        mAnimationHandler = new AnimationHandlerImpl(item, this);
        mFetcherHandler = new FetcherHandlerImpl(this, mAnimationHandler);
        mAnimationHandler.setFetcher(mFetcherHandler);
        return true;
    }
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        mFetcherHandler.onDestroy();
        mAnimationHandler.onDestroy();
    }
}

File: app/src/main/java/com/github/nickpakhomov/hotel_viewer/ui/activities/BaseActivity.kt
package com.github.nickpakhomov.hotel_viewer.ui.activities

import android.support.v7.app.AppCompatActivity
import android.view.Menu
import com.github.nickpakhomov.hotel_viewer.R
import com.github.nickpakhomov.hotel_viewer.handler.AnimationHandler
import com.github.nickpakhomov.hotel_viewer.handler.AnimationHandlerImpl
import com.github.nickpakhomov.hotel_viewer.handler.FetcherHandler
import com.github.nickpakhomov.hotel_viewer.handler.FetcherHandlerImpl

/**
 * Base activity that creates [FetcherHandlerImpl] and [AnimationHandlerImpl]
 * <br/>
 * Created by Nikolay Pakhomov on 08/09/17.
 */

open class BaseActivity : AppCompatActivity() {

    private lateinit var mAnimationHandler: AnimationHandler
    internal lateinit var mFetcherHandler: FetcherHandler

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (this is MainActivity)
            menuInflater.inflate(R.menu.menu_main, menu)
        else if (this is DetailActivity)
            menuInflater.inflate(R.menu.menu_detailed, menu)

        val item = menu.findItem(R.id.action_refresh)

        mAnimationHandler = AnimationHandlerImpl(item, this)
        mFetcherHandler = FetcherHandlerImpl(this, mAnimationHandler)
        mAnimationHandler.setFetcher(mFetcherHandler)
        return true
    }

    override fun onDestroy() {
        super.onDestroy()
        mFetcherHandler.onDestroy()
        mAnimationHandler.onDestroy()
    }
}
